mongo/change_stream.go

```diff
@@ -151,6 +151,26 @@ func mergeChangeStreamOptions(opts ...*options.ChangeStreamOptions) *options.Cha
 	return csOpts
 }
 
+// validChangeStreamTimeouts will return "false" if maxAwaitTimeMS is set,
+// timeoutMS is set to a non-zero value, and maxAwaitTimeMS is greater than or
+// equal to timeoutMS. Otherwise, the timeouts are valid.
+func validChangeStreamTimeouts(ctx context.Context, maxAwaitTime, timeout *time.Duration) bool {
```

Comment:
This validation does not make sense in the Go Driver. I believe it's based on [this requirement](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.md#tailable-awaitdata-cursors) from the CSOT spec:
> Drivers MUST also apply the original `timeoutMS` value to each next call on the resulting cursor

However, the Go Driver requires passing a `Context` to each call to `Next` or `TryNext`, so the timeout passed to the initial `Watch` has no bearing on those future `getMore` call timeouts.

We should remove this validation. However, we may want to apply an analogous validation when calling `Next` or `TryNext` if the `Context` timeouts for those are less than `maxAwaitTimeMS`.

Reply:
This validation is based on this requirement from the CSOT specs:

> Drivers MUST error if maxAwaitTimeMS is set, timeoutMS is set to a non-zero value, and maxAwaitTimeMS is greater than or equal to timeoutMS.

Reply:
Correct, that's the timeout validation that's implemented here. However, since the Go Driver doesn't do this:

> Drivers MUST also apply the original timeoutMS value to each next call on the resulting cursor

we also shouldn't do this:

> Drivers MUST error if maxAwaitTimeMS is set, timeoutMS is set to a non-zero value, and maxAwaitTimeMS is greater than or equal to timeoutMS.

because that validation only makes sense in drivers that use the timeout passed to `Find` as the timeout for `Next`. In the Go Driver, users pass separate timeouts to `Watch` and `Next`.

For example, the following code would cause a validation error even though there is no client-side timeout on `Next` (i.e. no conflict with `maxAwaitTimeMS`):
```go
opts := options.ChangeStream().SetMaxAwaitTime(10 * time.Second)

ctx, _ := context.WithTimeout(context.Background(), 1 * time.Second)
cs, err := coll.Watch(ctx, mongo.Pipeline{}, opts)
if err != nil {
	panic(err) // panics with "maxAwaitTimeMS cannot be greater than or equal to timeoutMS"
}

for cs.Next(context.Background()) {
	// ...
```

Additionally, the following code would _not_ cause a validation error, even though the `Next` timeout is much lower than `maxAwaitTimeMS`:
```go
opts := options.ChangeStream().SetMaxAwaitTime(10 * time.Second)

cs, err := coll.Watch(context.Background(), mongo.Pipeline{}, opts)
if err != nil {
	panic(err)
}

ctx, _ := context.WithTimeout(context.Background(), 1 * time.Second)
for cs.Next(ctx) {
	// Next possibly times out before a the server responds.
```

Reply:
> Drivers MUST also apply the original timeoutMS value to each next call on the resulting cursor

Doing this is part of the proposal for this PR, see [L712](https://github.com/mongodb/mongo-go-driver/pull/1515/files#diff-e167025d0c54c04fe112f526971955e98596621c3bdc99a4527311bc1025cf15R712).

Reply:
Based on additional discussion, I believe we decided to apply the `Watch` and `Next`/`TryNext` timeouts independently (both still using the client-level timeout when no operation-level timeout is set). As long as that's accurate, we can consider this comment resolved.